### PR TITLE
🔀 Support Distributed Tracing

### LINKED ISSUE #193: 🔀 Distributed Tracing
- [ ] Intake v2
- [ ] [test that responses from the server during chunked writes](https://github.com/elastic/apm-server/issues/1298#issuecomment-415704054) work with HTTP.rb
- [ ] `traceparent` header extract/inject functionality
- [ ] update opbeans-ruby to include a distributed tracing call
- [ ] `transaction.span_count.total` (https://github.com/elastic/apm-server/issues/1241)
- [ ] auto-instrument http module(s) to use extact/inject functionality

### PR DESCRIPTION
**Pull request by mikker**
Closes elastic/apm-agent-ruby#193

### CODE REVIEW

**Comment by mikker** [comment]
Another big one for when you have time, @simitt 😊

File: lib/elastic_apm/error_builder.rb [hunk 1]
```diff
@@ -19,10 +20,13 @@ def build_exception(exception, handled: true)
 
       if (transaction = ElasticAPM.current_transaction)
         error.context = transaction.context.dup
+        error.trace_id = transaction.trace_id
+        error.parent_id = transaction.id
```
Comment (simitt) [inline lib/elastic_apm/error_builder.rb]: The parent of an `error` could also be a `span`.  There's a dedicated `transaction_id` defined for an error, that should be set if available. See [json spec](https://github.com/elastic/apm-server/blob/v2/docs/spec/errors/v2_error.json#L21).

File: lib/elastic_apm/middleware.rb [hunk 2]
```diff
@@ -47,7 +50,15 @@ def path_ignored?(env)
...
       ElasticAPM.start_transaction 'Rack', 'request',
-        context: ElasticAPM.build_context(env)
+        context: ElasticAPM.build_context(env),
+        traceparent: traceparent(env)
+    end
+
+    def traceparent(env)
+      return unless (header = env['HTTP_ELASTIC_APM_TRACEPARENT'])
+      Traceparent.parse(header)
+    rescue Traceparent::InvalidTraceparentHeader
```
Comment (simitt) [inline lib/elastic_apm/middleware.rb]: Should you maybe log something here?

File: lib/elastic_apm/span.rb [hunk 3]
```diff
@@ -9,67 +9,75 @@ module ElasticAPM
...
-      @parent = parent
-      @context = context
 
       @id = SecureRandom.hex(8)
-      @trace_id = transaction.trace_id
-      @transaction_id = transaction.id
 
-      @timestamp = @transaction.timestamp
+      self.transaction = transaction
+      self.parent = parent
```
Comment (simitt) [inline lib/elastic_apm/span.rb]: From a consistency point of view - why using `self` here over `@`?

File: lib/elastic_apm/traceparent.rb [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    HEX_REGEX = /[^[:xdigit:]]/
+
+    def initialize
+      @version = VERSION
+    end
+
+    def self.from_transaction(transaction)
+      new.tap do |t|
+        t.trace_id = SecureRandom.hex(16)
+        t.span_id = transaction.id
```
Comment (simitt) [inline lib/elastic_apm/traceparent.rb]: Why do you set the `span_id` to `transaction.id`? From what I understand it should be set to the `span.id` that initiated the outgoing call to another service.

File: lib/elastic_apm/util.rb [hunk 5]
```diff
@@ -15,5 +15,13 @@ def self.git_sha
       sha = `git rev-parse --verify HEAD 2>&1`.chomp
       $? && $?.success? ? sha : nil # rubocop:disable Style/SpecialGlobalVars
     end
```
Comment (simitt) [inline lib/elastic_apm/util.rb]: I suggest to add the `.utc` inside the `def self.micros` method, to ensure all timestamps are used in utc.

File: spec/elastic_apm/middleware_spec.rb [hunk 6]
```diff
@@ -47,5 +47,77 @@ class MiddlewareTestError < StandardError; end
...
+
+      context 'with a blank header' do
+        it 'skips traceparent, makes new', :intercept do
+          app.call(
+            Rack::MockRequest.env_for('/', 'HTTP_ELASTIC_APM_TRACEPARENT' => '')
+          )
+
+          traceparent = @intercepted.transactions.first.traceparent
+          expect(traceparent.trace_id)
+            .to_not eq '0af7651916cd43dd8448eb211c80319c'
```
Comment (simitt) [inline spec/elastic_apm/middleware_spec.rb]: You could also check here that `traceparent.span_id` is empty.

File: lib/elastic_apm/span.rb (hunk 3 above)
Comment (simitt) [reply lib/elastic_apm/span.rb]: There is no custom setter for `parent`.

**Pull request merged**